lib/srv/reexec.go

```diff
@@ -740,63 +573,380 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
+	if err := os.Chdir(homeDir); err != nil {
+		return errorWriter, teleport.RemoteCommandFailure, trace.Wrap(err, "failed to set working directory for networking process")
+	}
+
+	// Build request listener from first extra file that was passed to command.
 	ffd := os.NewFile(FirstExtraFile, "listener")
 	if ffd == nil {
 		return errorWriter, teleport.RemoteCommandFailure, trace.BadParameter("missing socket fd")
 	}
 
-	conn, err := uds.FromFile(ffd)
+	parentConn, err := uds.FromFile(ffd)
 	ffd.Close()
 	if err != nil {
 		return errorWriter, teleport.RemoteCommandFailure, trace.Wrap(err)
 	}
 
-	ctx, cancel := context.WithCancel(context.Background())
+	// parentConn is a datagram Unix socket, which is not connection oriented
+	// and thus won't unblock when the parent closes its side of the connection.
+	// Instead we use an interrupt signal from the parent process to close this
+	// side of the connection and unblock the main request goroutine.
+	ctx, cancel := signal.NotifyContext(context.Background(), os.Interrupt)
 	defer cancel()
+	context.AfterFunc(ctx, func() { parentConn.Close() })
+
+	writeErrorToConn := func(conn io.Writer, err error) {
+		conn.Write([]byte(err.Error()))
+	}
+
+	// Some PAM modules (e.g. pam_namespace) do not behave properly in multithreaded contexts.
+	// Therefore we favor creating networking resources in the main PAM thread if we expect them
+	// to be impacted.
+	requestC := make(chan *networkingRequest)
 
+	// Create a waitgroup to ensure that the cleanup channel is closed only once
+	// all ongoing request connections are closed and their networking resources
+	// have been cleaned up or sent to the cleanup channel.
+	var wg sync.WaitGroup
+
+	// Networking resources created from in main PAM thread should also be cleaned up
+	// in the main PAM thread. Cleanup will either happen when the networking resources's
+	// connection to the parent is closed, or when this process exits (deferred).
+	cleanupC := make(chan networkingResource)
+	defer func() {
+		go func() {
+			wg.Wait()
+			close(cleanupC)
+		}()
+
+		for netResource := range cleanupC {
+			netResource.Close()
+		}
+	}()
+
+	// Start the main request goroutine.
+	go func() {
+		for {
+			buf := make([]byte, networking.RequestBufferSize)
+			fbuf := make([]*os.File, 1)
+			n, fn, err := uds.ReadWithFDs(parentConn, buf, fbuf)
+			if err != nil {
+				// parent connection closed by interrupt signal.
+				if utils.IsOKNetworkError(err) {
+					return
+				}
+				writeErrorToConn(parentConn, err)
+				continue
+			}
+
+			if fn == 0 {
+				writeErrorToConn(parentConn, trace.BadParameter("networking request missing per-request uds file"))
+				continue
+			}
+
+			req := new(networkingRequest)
+			req.conn, err = uds.FromFile(fbuf[0])
+			_ = fbuf[0].Close()
+			if err != nil {
+				writeErrorToConn(parentConn, trace.BadParameter("failed to get a connection from control file"))
+				continue
+			}
+			context.AfterFunc(ctx, func() { req.conn.Close() })
```

Comment:
This should be undone once we're done with `req.conn`, or it'll accumulate callbacks in the context forever.